### PR TITLE
Superstruct Validation Upgrade

### PR DESCRIPTION
**Pull request by pckilgore**
Right now certain aspects of testing this are disabled until ianstormtaylor/superstruct#122 or ianstormtaylor/superstruct#126 resolves.  There was some action today after I bumped it so I'll look this weekend because I want to get to know the lib better anyways.  

Ready to be merged now, however, we can enable those parts of the tests when we care (not essential to core features of validation).

### CODE REVIEW

File: src/utils/json-schema.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import _ from "underscore";
```
Comment (pckilgore) [inline src/utils/json-schema.js]: These functions were moved from `src/withForm/index.js` without modification (other than removing rjsf stuff).

File: test/fixtures/offering/test-06.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+const { testData: goodData } = require("./test-00-valid-data");
+
+module.exports = {
+  inputType: "UpdateDealInput",
+  isValid: false,
+  // message:
```
Comment (pckilgore) [inline test/fixtures/offering/test-06.js]: Can uncomment once library PR goes in (see original PR message).

File: src/validation/index.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+ *  Named after types, should be Capitalized.
+ */
+const AWSDate = value =>
+  handleRequired(value) || validISO8601.test(value) || "invalid_date";
+const AWSEmail = value =>
+  handleRequired(value) || isEmail(value) || "invalid_email";
+
+const AWSPhone = value => {
+  try {
+    const number = parsePhoneNumber(value);
```
Comment (ikeforrest) [inline src/validation/index.js]: why the need to assign this to a var?

File: src/validation/index.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    return number.isValid();
+  } catch (err) {
+    if (err instanceof ParseError) return "phone_" + err.message.toLowerCase();
```
Comment (ikeforrest) [inline src/validation/index.js]:
```suggestion
    if (err instanceof ParseError) return `phone_${err.message.toLowerCase()}`;
```

File: src/validation/index.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+ */
+const typeMatcher = ({ apiSchema, field, required = false }) => {
+  const match = field.type || field.ofType;
+
+  switch (match.kind) {
+    case "NON_NULL":
+      return typeMatcher({ apiSchema, field: match, required: true });
+
+    case "SCALAR":
+      return required ? match.name : Option(match.name);
```
Comment (ikeforrest) [inline src/validation/index.js]: for some reason ternaries are hard on the eyes here..IMO if/else would read easier here.

File: src/withForm/index.js [hunk 6]
```diff
@@ -120,73 +49,28 @@ export const withForm = ({
...
-      }
+    withHandlers(initialProps => {
+      // Expensive ops.  Do them once, keep in closure.
+      // API dict creation should really only happen once, or maybe on build.
+      const apiSchema = createApiDict(initialProps.apiSchema);
+      const validator = createValidator({ apiSchema, inputType: input });
+
+      return {
+        validateFormData: ({ setErrors }) => data => {
+          try {
```
Comment (ikeforrest) [inline src/withForm/index.js]: whitespace is your friend lol

File: test/fixtures/offering/test-10.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
+const { testData: goodData } = require("./test-00-valid-data");
+
+module.exports = {
+  inputType: "UpdateDealInput",
+  isValid: false,
+  // message:
+  //   'Expected a value of type `AWSJSON` for `sections.0.info.0.content.json` but received `"Unquoted String"`.',
+  testData: {
+    ...goodData,
```
Comment (ikeforrest) [inline test/fixtures/offering/test-10.js]: I about had a heart attack thinking this var name was in the code and not a test lol

File: src/validation/index.js (hunk 3 above)
Comment (pckilgore) [reply src/validation/index.js]: Method chaining with this library does weird shit.  It's not ideal but I'd rather do something like this than have to worry about a bunch of weird errors.  Even almost all their own docs do something similar.

File: src/validation/index.js (hunk 5 above)
Comment (pckilgore) [reply src/validation/index.js]: Hard disagree here.  IMO the power of a match expression comes with their terseness and flatness.  

One path in, one path out.  If/Else means multiple return statements, and adding two more scopes to each case.   It also feels weird wrapping a function around a language feature (ie `const typeOrOption = type => required ? type : Option(type)`)

File: src/withForm/index.js (hunk 6 above)
Comment (pckilgore) [reply src/withForm/index.js]: There are only four places prettier would let me, and none of them are helpful.

image

File: test/fixtures/offering/test-10.js (hunk 7 above)
Comment (pckilgore) [reply test/fixtures/offering/test-10.js]: You can't use qewl-forms?  You make sure to use the goodData?

File: src/withForm/index.js (hunk 6 above)
Comment (pckilgore) [reply src/withForm/index.js]: Five.  Whatever, blame prettier, I like whitespace.

**Comment by pckilgore** [comment]
fixes groundbreaker/tickets#686

**Pull request merged**